File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/TransferProcessFragment.java
package org.mifos.mobilebanking.ui.fragments;

import android.graphics.drawable.Animatable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatButton;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.models.payload.TransferPayload;
import org.mifos.mobilebanking.presenters.TransferProcessPresenter;
import org.mifos.mobilebanking.ui.activities.SavingsAccountContainerActivity;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.enums.TransferType;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.TransferProcessView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.CurrencyUtil;
import org.mifos.mobilebanking.utils.Network;
import org.mifos.mobilebanking.utils.Toaster;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by dilpreet on 1/7/17.
 */

public class TransferProcessFragment extends BaseFragment implements TransferProcessView {

    @BindView(R.id.tv_amount)
    TextView tvAmount;

    @BindView(R.id.tv_pay_to)
    TextView tvPayTo;

    @BindView(R.id.tv_pay_from)
    TextView tvPayFrom;

    @BindView(R.id.tv_date)
    TextView tvDate;

    @BindView(R.id.tv_remark)
    TextView tvRemark;

    @BindView(R.id.iv_success)
    ImageView ivSuccess;

    @BindView(R.id.ll_transfer)
    LinearLayout llTransfer;

    @BindView(R.id.btn_close)
    AppCompatButton btnClose;

    @Inject
    TransferProcessPresenter presenter;

    private View rootView;
    private TransferPayload payload;
    private TransferType transferType;

    /**
     * Used for TPT Transfer and own Account Transfer.<br>
     * Use {@code type} as TransferType.TPT for TPT and TransferType.SELF for self Account Transfer
     *
     * @param payload Transfer Information
     * @param type    enum of {@link TransferType}
     * @return Instance of {@link TransferProcessFragment}
     */
    public static TransferProcessFragment newInstance(TransferPayload payload, TransferType type) {
        TransferProcessFragment fragment = new TransferProcessFragment();
        Bundle args = new Bundle();
        args.putParcelable(Constants.PAYLOAD, payload);
        args.putSerializable(Constants.TRANSFER_TYPE, type);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getActivity() != null) {
            payload = getArguments().getParcelable(Constants.PAYLOAD);
            transferType = (TransferType) getArguments().getSerializable(Constants.TRANSFER_TYPE);
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_transfer_process, container, false);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        setToolbarTitle(getString(R.string.transfer));

        ButterKnife.bind(this, rootView);
        presenter.attachView(this);

        tvAmount.setText(CurrencyUtil.formatCurrency(getActivity(), payload.getTransferAmount()));
        tvPayFrom.setText(String.valueOf(payload.getFromAccountId()));
        tvPayTo.setText(String.valueOf(payload.getToAccountId()));
        tvDate.setText(payload.getTransferDate());
        tvRemark.setText(payload.getTransferDescription());

        return rootView;
    }

    /**
     * Initiates a transfer depending upon {@code transferType}
     */
    @OnClick(R.id.btn_start_transfer)
    public void startTransfer() {
        if (!Network.isConnected(getActivity())) {
            Toaster.show(rootView, getString(R.string.internet_not_connected));
            return;
        }
        if (transferType == TransferType.SELF) {
            presenter.makeSavingsTransfer(payload);
        } else if (transferType == TransferType.TPT) {
            presenter.makeTPTTransfer(payload);
        }
    }

    /**
     * Cancels the Transfer and pops fragment
     */
    @OnClick(R.id.btn_cancel_transfer)
    public void cancelTransfer() {
        Toaster.cancelTransfer(rootView, getString(R.string.cancel_transfer),
                getString(R.string.yes), new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        getActivity().getSupportFragmentManager().popBackStack();
                        getActivity().getSupportFragmentManager().popBackStack();
                    }
                });
    }

    /**
     * Closes the transfer fragment
     */
    @OnClick(R.id.btn_close)
    public void closeClicked() {
        getActivity().getSupportFragmentManager().popBackStack();
        getActivity().getSupportFragmentManager().popBackStack();
    }

    /**
     * Shows a {@link android.support.design.widget.Snackbar} on succesfull transfer of money
     */
    @Override
    public void showTransferredSuccessfully() {
        Toaster.show(rootView, getString(R.string.transferred_successfully));
        ivSuccess.setVisibility(View.VISIBLE);
        ((Animatable) ivSuccess.getDrawable()).start();
        btnClose.setVisibility(View.VISIBLE);
        llTransfer.setVisibility(View.GONE);
        SavingsAccountContainerActivity.transferSuccess = true;
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param msg Error message that tells the user about the problem.
     */
    @Override
    public void showError(String msg) {
        Toaster.show(rootView, msg);
    }

    @Override
    public void showProgress() {
        showMifosProgressDialog(getString(R.string.please_wait));
    }

    @Override
    public void hideProgress() {
        hideMifosProgressDialog();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        presenter.detachView();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/TransferProcessFragment.kt
package org.mifos.mobile.ui.fragments

import android.graphics.drawable.Animatable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentTransferProcessBinding
import org.mifos.mobile.models.payload.TransferPayload
import org.mifos.mobile.models.templates.account.AccountOption
import org.mifos.mobile.ui.activities.SavingsAccountContainerActivity
import org.mifos.mobile.ui.enums.TransferType
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.*
import org.mifos.mobile.viewModels.TransferProcessViewModel

/**
 * Created by dilpreet on 1/7/17.
 */
@AndroidEntryPoint
class TransferProcessFragment : BaseFragment() {

    private var _binding: FragmentTransferProcessBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel: TransferProcessViewModel

    private var toAccountOption: AccountOption? = null
    private var fromAccountOption: AccountOption? = null

    private var payload: TransferPayload? = null
    private var transferType: TransferType? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (activity != null) {
            payload = arguments?.getParcelable(Constants.PAYLOAD)
            transferType = arguments?.getSerializable(Constants.TRANSFER_TYPE) as TransferType
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentTransferProcessBinding.inflate(inflater, container, false)
        setToolbarTitle(getString(R.string.transfer))
        viewModel = ViewModelProvider(this)[TransferProcessViewModel::class.java]
        with(binding) {
            tvAmount.text = CurrencyUtil.formatCurrency(activity, payload?.transferAmount)
            tvPayFrom.text = payload?.fromAccountNumber.toString()
            tvPayTo.text = payload?.toAccountNumber.toString()
            tvDate.text = payload?.transferDate
            tvRemark.text = payload?.transferDescription
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.transferUiState.observe(viewLifecycleOwner) {
            when (it) {
                is TransferUiState.Loading -> showProgress()
                is TransferUiState.TransferSuccess -> {
                    hideProgress()
                    showTransferredSuccessfully()
                }
                is TransferUiState.Error -> {
                    hideProgress()
                    showError(MFErrorParser.errorMessage(it.errorMessage))
                }
            }
        }

        with(binding) {
            btnStartTransfer.setOnClickListener {
                startTransfer()
            }
            btnCancelTransfer.setOnClickListener {
                cancelTransferProcess()
            }
            btnClose.setOnClickListener {
                closeClicked()
            }
        }
    }

    /**
     * Initiates a transfer depending upon `transferType`
     */
    private fun startTransfer() {
        if (!Network.isConnected(activity)) {
            Toaster.show(binding.root, getString(R.string.internet_not_connected))
            return
        }
        viewModel.makeTransfer(
            fromAccountOption?.officeId,
            fromAccountOption?.clientId,
            fromAccountOption?.accountType?.id,
            fromAccountOption?.accountId,
            toAccountOption?.officeId,
            toAccountOption?.clientId,
            toAccountOption?.accountType?.id,
            toAccountOption?.accountId,
            DateHelper.getSpecificFormat(
                DateHelper.FORMAT_dd_MMMM_yyyy,
                getTodayFormatted(),
            ),
            binding.tvAmount.text.toString().toDouble(),
            binding.tvRemark.text.toString(),
            "dd MMMM yyyy",
            "en",
            fromAccountOption?.accountNo,
            toAccountOption?.accountNo,
            transferType
        )

    }

    /**
     * Cancels the Transfer and pops fragment
     */
    fun cancelTransferProcess() {
        Toaster.cancelTransfer(
            binding.root,
            getString(R.string.cancel_transfer),
            getString(R.string.yes),
            View.OnClickListener {
                activity?.supportFragmentManager?.popBackStack()
                activity?.supportFragmentManager?.popBackStack()
            },
        )
    }

    /**
     * Closes the transfer fragment
     */
    private fun closeClicked() {
        activity?.supportFragmentManager?.popBackStack()
        activity?.supportFragmentManager?.popBackStack()
    }

    /**
     * Shows a {@link Snackbar} on succesfull transfer of money
     */
    fun showTransferredSuccessfully() {
        Toaster.show(binding.root, getString(R.string.transferred_successfully))
        binding.ivSuccess.visibility = View.VISIBLE
        (binding.ivSuccess.drawable as Animatable).start()
        binding.btnClose.visibility = View.VISIBLE
        binding.llTransfer.visibility = View.GONE
        SavingsAccountContainerActivity.transferSuccess = true
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param msg Error message that tells the user about the problem.
     */
    fun showError(msg: String?) {
        Toaster.show(binding.root, msg)
    }

    fun showProgress() {
        showMifosProgressDialog(getString(R.string.please_wait))
    }

    fun hideProgress() {
        hideMifosProgressDialog()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        /**
         * Used for TPT Transfer and own Account Transfer.<br></br>
         * Use `type` as TransferType.TPT for TPT and TransferType.SELF for self Account Transfer
         *
         * @param payload Transfer Information
         * @param type    enum of [TransferType]
         * @return Instance of [TransferProcessFragment]
         */
        fun newInstance(payload: TransferPayload?, type: TransferType?): TransferProcessFragment {
            val fragment = TransferProcessFragment()
            val args = Bundle()
            args.putParcelable(Constants.PAYLOAD, payload)
            args.putSerializable(Constants.TRANSFER_TYPE, type)
            fragment.arguments = args
            return fragment
        }
    }
}
